### PR TITLE
feat(dgw): streaming recordings

### PR DESCRIPTION
**Pull request by irvingoujAtDevolution**
# Pull Request for Discussion: Real-time Recording Streaming Endpoints

## Summary

This PR introduces two new API endpoints:

- `/jet/jrec/list/realtime`:  
  Lists all WebM recordings currently being streamed in real time.

- `/jet/jrec/realtime/:id/:filename`:  
  A WebSocket endpoint for streaming an active WebM recording session.

Additionally, a new crate `streamer` has been added to encapsulate the logic for streaming WebM files.

## Streaming Process

1. **User initiates a WebM recording session.**
   
2. **Server begins recording** and registers a signal with the `RecordingMessageSender`. Each time new data is appended to the file and a flush occurs, a message is emitted.  
    - Note: This approach might not fully align with the semantics of `RecordingMessageSender` because it breaks its intended message flow. Feedback on this aspect would be appreciated.

3. **User connects to the WebSocket endpoint** at `/jet/jrec/realtime/:id/:filename`.

4. **Server reads the file header** of the ongoing recording and identifies the offset of the latest cluster start.

5. **Server sends the necessary metadata** to the client.

6. **Server streams the recording in real time**:
    - The server continuously reads from the `last_cluster_position` and, if the data from this position to the end of the file can be parsed into a complete cluster, it is sent to the client, with updated timestamp.
    - If the cluster is incomplete, the server waits for new data to be appended (triggered by the `RecordingMessageSender`) and retries sending the full cluster once available.

## Discussion

This is a functional version of the feature, built on top of the previously submitted PR. I’d love to hear your thoughts, especially regarding potential performance concerns or any other improvements you’d suggest.

@CBenoit, please feel free to share your input so we can refine this further.

p.s we can skip the details for this one. :) since it will likely not be merged like this.

### CODE REVIEW

File: devolutions-gateway/src/api/jrec.rs [hunk 1]
```diff
@@ -25,6 +25,8 @@ pub fn make_router<S>(state: DgwState) -> Router<S> {
         .route("/delete/:id", delete(jrec_delete))
         .route("/list", get(list_recordings))
         .route("/pull/:id/:filename", get(pull_recording_file))
+        .route("/list/realtime", get(list_realtime_recordings))
+        .route("/realtime/:id/:filename", get(stream_recording))
```
Comment (CBenoit) [inline devolutions-gateway/src/api/jrec.rs:29]: suggestion:
- `GET /jet/jrec/list-active`
- `GET /shadow/:id`

File: devolutions-gateway/src/api/webapp.rs [hunk 2]
```diff
@@ -285,6 +285,7 @@ pub(crate) struct SessionTokenSignRequest {
         ("web_app_token" = []),
     ),
 ))]
+
```
Comment (CBenoit) [inline devolutions-gateway/src/api/webapp.rs:288]: issue: Double check the PR to remove this kind of unintentional changes.

File: devolutions-gateway/src/streaming/mod.rs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    recording_id: Uuid,
+) -> anyhow::Result<Response<Body>> {
+    // 1.identify the file type
+    if !(path.extension() == Some(RecordingFileType::WebM.extension())) {
+        return Err(anyhow::anyhow!("Invalid file type"));
+    }
+    // 2.if the file is actively being recorded, then proceed
+    let Ok(Some(OnGoingRecordingState::Connected)) = recordings.get_state(recording_id).await else {
+        return Err(anyhow::anyhow!("File is not being recorded"));
+    };
```
Comment (CBenoit) [inline devolutions-gateway/src/streaming/mod.rs:24]: praise: Are you trying to follow some [literate programming-like](https://en.wikipedia.org/wiki/Literate_programming) convention?

style: https://github.com/Devolutions/IronRDP/blob/87014d4afb3229813f4a61ab30fc7afd778e3fb4/STYLE.md#inline-code-comments-are-proper-sentences

File: devolutions-gateway/src/streaming/mod.rs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+    let streaming_file = ReOpenableFile::open(&path).with_context(|| format!("Failed to open file: {:?}", path))?;
```
Comment (CBenoit) [inline devolutions-gateway/src/streaming/mod.rs:26]: style: https://github.com/Devolutions/IronRDP/blob/87014d4afb3229813f4a61ab30fc7afd778e3fb4/STYLE.md#formatting-of-error-messages
```suggestion
    let streaming_file = ReOpenableFile::open(&path).with_context(|| format!("failed to open file: {path:?}"))?;
```

File: devolutions-gateway/src/streaming/mod.rs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+            })
+            .with_context(|| format!("webm_stream failed"))?;
+            Ok::<_, anyhow::Error>(())
+        })
+        .await;
+
+        let res = match streaming_result {
+            Ok(res) => res,
+            Err(e) => {
+                error!("Error while streaming file on join: {:?}", e);
```
Comment (CBenoit) [inline devolutions-gateway/src/streaming/mod.rs:47]: style: https://github.com/Devolutions/IronRDP/blob/87014d4afb3229813f4a61ab30fc7afd778e3fb4/STYLE.md#logging

**Review by CBenoit — COMMENTED** [review]
I only reviewed the HTTP API for now.

Potential issue I see currently: the endpoint `GET /jet/jrec/realtime/:id/:filename` (that I suggested to change to `GET /jet/jrec/shadow/:id`) is expecting a filename, but a recording may be cut into several files. This is fine for the prototype, but we should consider what to do when the recording stream stops, and starts again with a new file (which may have a different resolution).

This leads me to the next question: don’t we want a route akin to `GET /jet/jrec/player`, serving the shadowing player? Most likely, we want the shadowing player to handle the previous issue in some form.

**Comment by irvingoujAtDevolution** [comment]
> I only reviewed the HTTP API for now.
> 
> Potential issue I see currently: the endpoint `GET /jet/jrec/realtime/:id/:filename` (that I suggested to change to `GET /jet/jrec/shadow/:id`) is expecting a filename, but a recording may be cut into several files. This is fine for the prototype, but we should consider what to do when the recording stream stops, and starts again with a new file (which may have a different resolution).
> 
> This leads me to the next question: don’t we want a route akin to `GET /jet/jrec/player`, serving the shadowing player? Most likely, we want the shadowing player to handle the previous issue in some form.

I intent to use it with current recording player, if it is being streaming, then we could simply add another query parameters.

**Comment by irvingoujAtDevolution** [comment]
>Potential issue I see currently: the endpoint GET /jet/jrec/realtime/:id/:filename (that I suggested to change to GET /jet/jrec/shadow/:id) is expecting a filename, but a recording may be cut into several files. This is fine for the prototype, but we should consider what to do when the recording stream stops, and starts again with a new file (which may have a different resolution).

This rings a bell to me that we don't have any mechanism to deal with resolution or size change in recording or recording player at this time.

File: devolutions-gateway/src/recording.rs [hunk 6]
```diff
@@ -285,6 +311,28 @@ impl RecordingMessageSender {
...
             .context("couldn't send UpdateRecordingPolicy message")
     }
+
+    pub(crate) fn on_new_chunk_appended(&self, recording_id: Uuid, tx: oneshot::Sender<()>) -> anyhow::Result<()> {
+        let mut lock = self.flush_map.lock();
+        let senders = lock.entry(recording_id);
+        let senders = senders.or_default();
+        senders.push(tx);
+        Ok(())
+    }
```
Comment (CBenoit) [inline devolutions-gateway/src/recording.rs:321]: naming: Intent is not clear. I suggest something like `add_new_chunk_listener` or something like that.

**Review by CBenoit — COMMENTED** [review]
> edit 2nd time: How about let's wait until the RDM and web recording side agrees on how to handle separate files?

Yeah, it would be easier for us to test the shadowing player for this case once they actually do that.
Do you have plan for looking into that on web clients?
I think the canvas change its size via dynamic resizing for IronRDP, do you know what happens in such cases?

That being said, I suggest we still go ahead and merge your initial implementation sooner than later, but put the new endpoints behind the unstable debug flag.

https://github.com/Devolutions/devolutions-gateway/blob/5169f6062231b92b87236ec38b6987f18659dfdf/devolutions-gateway/src/api/net.rs#L16-L21

Several benefits:
- As long as things are behind this flag, we can change without caring about breaking the API boundary.
- We don’t accumulate conflicts because changes are in the main source tree.
- It’s possible to ship prerelease and release binaries with the unstable feature and test the experimental features internally without requiring the Rust toolchain for everyone.

File: devolutions-gateway/src/recording.rs [hunk 7]
```diff
@@ -107,26 +108,45 @@ where
...
 
+        debug!(path = %recording_file, "File opened");
+
         let res = match open_options.open(&recording_file).await {
             Ok(file) => {
-                let mut file = BufWriter::new(file);
+                // Wrap SignalWriter inside a BufWriter to reduce the number of flushes
+                let (file, mut flush_signal) = SignalWriter::new(file);
+                // larger buffer size to reduce the number of flushes
+                let mut file = BufWriter::with_capacity(1024 * 100, file);
```
[the remark this answers is not in the document]
Comment (irvingoujAtDevolution) [reply devolutions-gateway/src/recording.rs]: Updated, each cluster to approximately 500,000 bytes . by setting the buffer around 1/5 of it, we avoid holding too much data in the buffer as the clusters stream in, and we can parse them promptly without excessive buildup.

[the remark this answers is not in the document]
Comment (irvingoujAtDevolution) [reply devolutions-gateway/src/recording.rs]: but I updated it to 64 * 1024, we can always adjust it later if necessary.

**Comment by CBenoit** [comment]
I’ll perform another review pass and merge by tomorrow!

**Comment by irvingoujAtDevolution** [comment]
> I’ll perform another review pass and merge by tomorrow!

Hi @CBenoit , is there anything else should I update here?

**Comment by CBenoit** [comment]
@irvingoujAtDevolution I was about to merge. I see you push-forced the branch, did you change anything?

File: tools/tokengen/Cargo.toml [hunk 8]
```diff
@@ -4,25 +4,16 @@ version = "0.0.0"
...
 serde = "1.0"
 serde_json = "1.0"
 uuid = { version = "1.1", features = ["v4", "serde"] }
 tap = "1.0"
-tokio = { version = "1.41", features = ["fs", "io-util", "net", "rt", "rt-multi-thread"] }
-axum = { version = "0.7" }
-tower = { version = "0.5" }
-tracing = { version = "0.1" }
-tracing-subscriber = { version = "0.3", features = ["env-filter"] }
-tower-http = { version = "0.6.1", features = ["cors", "trace"] }
```
Comment (CBenoit) [inline tools/tokengen/Cargo.toml:28]: Did you intend to remove that?

File: devolutions-gateway/src/api/jrec.rs [hunk 9]
```diff
@@ -20,14 +20,22 @@ use crate::token::{JrecTokenClaims, RecordingFileType, RecordingOperation};
...
         .route("/list", get(list_recordings))
         .route("/pull/:id/:filename", get(pull_recording_file))
         .route("/play", get(get_player))
-        .route("/play/*path", get(get_player))
-        .with_state(state)
+        .route("/play/*path", get(get_player));
+
+    if state.conf_handle.get_conf().debug.enable_unstable {
+        router
+            .route("/list-active", get(list_active_recordings))
```
Comment (CBenoit) [inline devolutions-gateway/src/api/jrec.rs:33]: suggestion: I like François' idea. Let's modify `/list` to take a query param such as "?active=true/false".

File: jetsocat/Cargo.toml [hunk 10]
```diff
@@ -36,7 +36,7 @@ seahorse = "2.2"
 humantime = "2.1"
 
 # async
-tokio = { version = "1.41", features = ["io-std", "io-util", "net", "fs", "signal", "time", "rt", "sync", "process", "rt-multi-thread", "macros"] }
+tokio = { version = "1.38", features = ["io-std", "io-util", "net", "fs", "signal", "time", "rt", "sync", "process", "rt-multi-thread", "macros"] }
```
Comment (CBenoit) [inline jetsocat/Cargo.toml:39]: You probably got a problem when you merged changes from master, I see some things being reverted here

File: devolutions-gateway/src/api/jrec.rs (hunk 9 above)
Comment (irvingoujAtDevolution) [reply devolutions-gateway/src/api/jrec.rs]: Updated

File: jetsocat/Cargo.toml (hunk 10 above)
Comment (irvingoujAtDevolution) [reply jetsocat/Cargo.toml]: updated

File: tools/tokengen/Cargo.toml (hunk 8 above)
Comment (irvingoujAtDevolution) [reply tools/tokengen/Cargo.toml]: updated

**Comment by CBenoit** [comment]
I’ll likely review tomorrow.

File: tools/tokengen/src/server/server_impl.rs [hunk 11]
```diff
@@ -56,7 +56,7 @@ pub(crate) async fn get_provisioner_key_path() -> Result<Arc<PathBuf>, Box<dyn E
 
     let gateway_json_path = Path::new(&config_dir).join("gateway.json");
 
-    let gateway_json_contents = tokio::fs::read_to_string(&gateway_json_path).await?;
+    let gateway_json_contents:String = tokio::fs::read_to_string(&gateway_json_path).await?;
```
Comment (CBenoit) [inline tools/tokengen/src/server/server_impl.rs:59]: style: Run cargo fmt.

File: tools/tokengen/src/server/server_impl.rs [hunk 12]
```diff
@@ -56,7 +56,7 @@ pub(crate) async fn get_provisioner_key_path() -> Result<Arc<PathBuf>, Box<dyn E
...
     let gateway_config: serde_json::Value = serde_json::from_str(&gateway_json_contents)?;
```
Comment (CBenoit) [inline tools/tokengen/src/server/server_impl.rs:60]: suggestion: You need to run `cargo fmt` on `tokengen`, but I don’t think this change is necessary in the first place as `read_to_string` will obviously return a `String`.

File: devolutions-gateway/src/recording.rs [hunk 13]
```diff
@@ -150,8 +178,8 @@ impl ActiveRecordings {
         self.0.lock().contains(&id)
     }
 
-    /// Returns a copy of the internal HashSet
-    pub fn cloned(&self) -> HashSet<Uuid> {
+    /// Get the list of active recordings without locking the mutex
+    pub fn copy_set(&self) -> HashSet<Uuid> {
         self.0.lock().clone()
     }
```
Comment (CBenoit) [inline devolutions-gateway/src/recording.rs:184]: It seems you overrode code already present. Revert these changes.

File: devolutions-gateway/src/api/jrec.rs [hunk 14]
```diff
@@ -245,9 +255,9 @@ async fn jrec_delete_many(
     fn process_request(
         delete_list: Vec<Uuid>,
         recording_path: &Utf8Path,
-        active_recordings: &HashSet<Uuid>,
+        active_recordings: &ActiveRecordings,
     ) -> Result<ProcessResult, HttpError> {
-        let conflict = delete_list.iter().any(|id| active_recordings.contains(id));
+        let conflict = delete_list.iter().any(|id| active_recordings.contains(*id));
```
Comment (CBenoit) [inline devolutions-gateway/src/api/jrec.rs:260]: Revert this too. It’s much less efficient as this version will need to acquire the lock for each value in the delete list.

File: devolutions-gateway/src/api/jrec.rs [hunk 15]
```diff
@@ -38,6 +44,12 @@ struct JrecPushQueryParam {
     file_type: RecordingFileType,
 }
 
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub(crate) struct JrecListQueryParam {
+    active: Option<bool>,
```
Comment (CBenoit) [inline devolutions-gateway/src/api/jrec.rs:50]: suggestion: Use a `bool` which will be false by default instead of an `Option<bool>`

File: devolutions-gateway/src/api/jrec.rs [hunk 16]
```diff
@@ -376,6 +397,11 @@ pub(crate) async fn list_recordings(
     }
 }
 
+async fn list_active_recordings(recordings: RecordingMessageSender) -> Result<Json<Vec<Uuid>>, HttpError> {
+    let recordings = recordings.active_recordings.copy_set().into_iter().collect();
+    Ok(Json(recordings))
+}
```
Comment (CBenoit) [inline devolutions-gateway/src/api/jrec.rs:403]: issue: This function is not performing any async operation.

suggestion: Inline the function as it’s reused nowhere and is only a very readable single liner.

File: tools/tokengen/src/server/server_impl.rs (hunk 12 above)
Comment (irvingoujAtDevolution) [reply tools/tokengen/src/server/server_impl.rs]: updated

File: devolutions-gateway/src/api/jrec.rs (hunk 15 above)
Comment (irvingoujAtDevolution) [reply devolutions-gateway/src/api/jrec.rs:50]: updated

File: devolutions-gateway/src/recording.rs (hunk 13 above)
Comment (irvingoujAtDevolution) [reply devolutions-gateway/src/recording.rs]: updated

File: devolutions-gateway/src/api/jrec.rs (hunk 14 above)
Comment (irvingoujAtDevolution) [reply devolutions-gateway/src/api/jrec.rs]: updated

File: devolutions-gateway/src/api/jrec.rs (hunk 16 above)
Comment (irvingoujAtDevolution) [reply devolutions-gateway/src/api/jrec.rs]: updated

File: crates/streamer/src/streamer/tag_writers.rs [hunk 17]
```diff
@@ -332,9 +341,13 @@ impl TryFrom<(Headers<'_>, &StreamingConfig)> for EncodeWriterConfig {
...
-            width: width.ok_or(anyhow::anyhow!("No width specified"))?,
-            height: height.ok_or(anyhow::anyhow!("No height specified"))?,
-            codec: codec.ok_or(anyhow::anyhow!("No codec specified"))?,
+            width: width
+                .map(u32::try_from)
+                .ok_or_else(|| anyhow::anyhow!("No width specified"))??,
+            height: height
+                .map(u32::try_from)
+                .ok_or_else(|| anyhow::anyhow!("No height specified"))??,
+            codec: codec.ok_or_else(|| anyhow::anyhow!("No codec specified"))?,
```
Comment (CBenoit) [inline crates/streamer/src/streamer/tag_writers.rs:353]: suggestion: You can use `Context`.

Comment (irvingoujAtDevolution) [reply crates/streamer/src/streamer/tag_writers.rs:353]: updated

File: crates/streamer/src/streamer/tag_writers.rs [hunk 18]
```diff
@@ -344,13 +344,9 @@ impl TryFrom<(Headers<'_>, &StreamingConfig)> for EncodeWriterConfig {
...
-            width: width
-                .map(u32::try_from)
-                .ok_or_else(|| anyhow::anyhow!("No width specified"))??,
-            height: height
-                .map(u32::try_from)
-                .ok_or_else(|| anyhow::anyhow!("No height specified"))??,
-            codec: codec.ok_or_else(|| anyhow::anyhow!("No codec specified"))?,
+            width: width.map(u32::try_from).context("No width specified")??,
+            height: height.map(u32::try_from).context("No height specified")??,
+            codec: codec.context("No codec specified")?,
```
Comment (CBenoit) [inline crates/streamer/src/streamer/tag_writers.rs:349]: https://github.com/Devolutions/IronRDP/blob/eb499de352d7da03f17dcd10f61c764c0bdeb81d/STYLE.md#formatting-of-error-messages

Comment (irvingoujAtDevolution) [reply crates/streamer/src/streamer/tag_writers.rs]: updated

**Comment by irvingoujAtDevolution** [comment]
I will remain active for new comments @CBenoit

File: devolutions-gateway/src/streaming/mod.rs
Comment (CBenoit) [inline devolutions-gateway/src/streaming/mod.rs:1]: Since there are no submodules, convert to `streaming.rs`

File: devolutions-gateway/src/streaming/mod.rs [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+pub(crate) async fn stream_file(
+    path: camino::Utf8PathBuf,
+    ws: axum::extract::WebSocketUpgrade,
+    shutdown_signal: devolutions_gateway_task::ShutdownSignal,
+    recordings: crate::recording::RecordingMessageSender,
+    recording_id: Uuid,
+) -> anyhow::Result<Response<Body>> {
+    // 1.identify the file type
+    if path.extension() != Some(RecordingFileType::WebM.extension()) {
+        return Err(anyhow::anyhow!("Invalid file type"));
```
Comment (CBenoit) [inline devolutions-gateway/src/streaming/mod.rs:25]: This is a nitpick that applies to many place in your PR: https://github.com/Devolutions/IronRDP/blob/fe0d9e9773d96ddb7b51c0c4e0d42821d0da569b/STYLE.md#formatting-of-error-messages
If you have time to address it do it, otherwise can wait

File: devolutions-gateway/src/streaming/mod.rs [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+        .await;
+
+        match streaming_result {
+            Err(e) => {
+                error!(?e, "streaming file task join failed");
+            }
+            Ok(Err(e)) => {
+                error!(?e, "streaming file failed");
+            }
+            _ => {}
```
Comment (CBenoit) [inline devolutions-gateway/src/streaming/mod.rs:64]: https://github.com/Devolutions/IronRDP/blob/fe0d9e9773d96ddb7b51c0c4e0d42821d0da569b/STYLE.md#logging

File: devolutions-gateway/src/streaming/mod.rs [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+                websocket_stream,
+                streaming_file,
+                ShutdownSignal(shutdown_signal),
+                // Question: Do we need to make it configurable
+                streamer::StreamingConfig { encoder_threads: 4 },
+                move || {
+                    let (tx, rx) = tokio::sync::oneshot::channel();
+                    recordings
+                        .add_new_chunk_listener(recording_id, tx)
+                        .expect("Failed to send on_appended message"); // early development
```
Comment (CBenoit) [inline devolutions-gateway/src/streaming/mod.rs:48]: Can you double check this comment? Should expect be removed?

File: devolutions-gateway/src/streaming/mod.rs [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+    let upgrade_result = ws.on_upgrade(move |socket| async move {
+        let websocket_stream = websocket_compat(socket);
+        // Spawn blocking because webm_stream is blocking
+        let streaming_result = tokio::task::spawn_blocking(move || {
+            webm_stream(
+                websocket_stream,
+                streaming_file,
+                ShutdownSignal(shutdown_signal),
+                // Question: Do we need to make it configurable
+                streamer::StreamingConfig { encoder_threads: 4 },
```
Comment (CBenoit) [inline devolutions-gateway/src/streaming/mod.rs:43]: We may wait to check for the number of cores instead of hardcoding 4, but requires some measures.

**Review by CBenoit — COMMENTED** [review]
We also need to:
- Publish a new version of the cadeau crate
- Update the DLL in download-cadeau.ps1

I didn’t get a chance to do [what I said here](https://github.com/Devolutions/cadeau/pull/55#pullrequestreview-2447252531).
If you have a Linux machine, maybe you can try on your side?
I’ll publish the new version of the cadeau crate after.

File: devolutions-gateway/src/streaming/mod.rs (hunk 22 above)
Comment (irvingoujAtDevolution) [reply devolutions-gateway/src/streaming/mod.rs]: updated!

File: devolutions-gateway/src/streaming/mod.rs (hunk 20 above)
Comment (irvingoujAtDevolution) [reply devolutions-gateway/src/streaming/mod.rs]: updated

File: devolutions-gateway/src/streaming/mod.rs (hunk 21 above)
Comment (irvingoujAtDevolution) [reply devolutions-gateway/src/streaming/mod.rs]: updated

File: devolutions-gateway/src/streaming/mod.rs (hunk 19 above)
Comment (irvingoujAtDevolution) [reply devolutions-gateway/src/streaming/mod.rs]: I checked with ripgrep, should be all fixed

File: devolutions-gateway/src/streaming/mod.rs
Comment (irvingoujAtDevolution) [reply devolutions-gateway/src/streaming/mod.rs:1]: updated

File: devolutions-gateway/src/api/jrec.rs [hunk 23]
```diff
@@ -344,6 +344,9 @@ async fn delete_recording(recording_path: &Utf8Path) -> anyhow::Result<()> {
     operation_id = "ListRecordings",
     tag = "Jrec",
     path = "/jet/jrec/list",
+    params(
+        ("active" = bool, Query, description = "When false, return only finished recordings, when true return only active recordings"),
```
Comment (CBenoit) [inline devolutions-gateway/src/api/jrec.rs:348]: When false, _all_ the recordings are returned. I’ll update the PR on my side.

**Review by CBenoit — APPROVED** [review]
LGTM! Merging as the initial implementation!
I set the https://github.com/Devolutions/devolutions-gateway/labels/release-blocker label because I want to follow up on the logic in the jrec module before cutting a new release.
